### PR TITLE
multi: add notifywork websocket rpc.

### LINKED ISSUE #1032: GetWork notifications over Websocket
Continuously polling `getwork` for changes (over HTTP or Websocket) is not really desirable and is the only way to get new work for miners at the moment.

I propose having a `notifywork` method on the websocket RPC, similar to [notifyblocks](https://github.com/decred/dcrd/blob/master/docs/json_rpc_api.md#notifyblocks). This starts a subscription on messages that could look something like this (suggestion):
```
{
	"jsonrpc": "1.0",
	"method": "latestwork",
	"params": [{
		"data": "block_header",
		"target": "some_target"
	}]
}
```

The object in the `params` list here is just the object returned from the regular `getwork`.

### PR DESCRIPTION
**Pull request by dnldd**
resolves #1032.

### CODE REVIEW

**Comment by dajohi** [comment]
@dnldd please rebase.

File: rpcclient/infrastructure.go [hunk 1]
```diff
@@ -545,6 +545,14 @@ func (c *Client) reregisterNtfns() error {
 		}
 	}
 
+	// Reregister notifyblocks if needed.
```
Comment (davecgh) [inline rpcclient/infrastructure.go]: notifywork

File: rpcclient/notify.go [hunk 2]
```diff
@@ -251,7 +258,23 @@ func (c *Client) handleNotification(ntfn *rawNotification) {
 
 		c.ntfnHandlers.OnBlockDisconnected(blockHeader)
 
-	// OnRelevantTxAccepted
+	// OnWork
+	case chainjson.WorkNtfnMethod:
+		// Ignore the notification if the client is not interested in
+		// it.
+		if c.ntfnHandlers.OnWork == nil {
+			log.Debugf("on work is nil")
```
Comment (davecgh) [inline rpcclient/notify.go]: It looks like this is leftover debug that should be removed.

File: rpcserver.go [hunk 3]
```diff
@@ -283,6 +283,7 @@ var rpcUnimplemented = map[string]struct{}{
 var rpcLimited = map[string]struct{}{
 	// Websockets commands
 	"notifyblocks":          {},
+	"notifywork":            {},
```
Comment (davecgh) [inline rpcserver.go]: I don't think this should available to limited since `getwork` isn't.

File: rpcwebsocket.go [hunk 4]
```diff
@@ -821,6 +864,64 @@ func (*wsNotificationManager) notifyBlockDisconnected(clients map[chan struct{}]
 	}
 }
 
+// notifyWork notifies websocket clients that have registered for template
+// updates when a new block template is generated.
+func (*wsNotificationManager) notifyWork(clients map[chan struct{}]*wsClient, header *wire.BlockHeader) {
+	// Skip notification creation if no clients have requested block
+	// connected/disconnected notifications.
```
Comment (davecgh) [inline rpcwebsocket.go]: work notifications

**Pull request merged**